Source/SGDLib/SimpleDistGradAggregator.h

```diff
@@ -326,36 +305,105 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
+            // CurrentGradientIndex will load the index from m_gradientIndexToAggregate
+            size_t currentGradientIndex = -2;  
+            //pipeline loop
             for (size_t i : m_gradientIndexToAggregate)
             {
-                allReduceRequests.push_back(MPI_Request());
-                reductionBuffer = (i == -1)? m_aggregationBuffer->Data() : gradients[i]->Data();
-                if (m_mpi->UseGpuGdr() == 0 && deviceId != CPUDEVICE)
+                if (i != -1)
                 {
-                    m_gpuDataTransferers[allReduceIndex]->WaitForCopyGPUToCPUAsync();
-                    reductionBuffer = m_intermediateCPUBuffers[allReduceIndex].get();
+                    gpuCopyBuffer = gradients[i];
+                }
+                else
+                {
+                    // i == -1, first element is for packed gradients, which should not be with AsyncAggregation
+                    assert(m_useAsyncAggregation == false);
+                }
+                //1.first sync_g_to_c_copy
+                if (gpuToCpuIndex == 0)
+                {
+                    //TODO: we need a CopyGPUToCPUSync
+                    cudaMemcpy(m_intermediateCPUBuffers[gpuToCpuIndex].get(), gpuCopyBuffer->Data(), gpuCopyBuffer->GetNumElements() * sizeof(ElemType), cudaMemcpyDeviceToHost);
+                    currentGradientIndex = i;
+                    gpuToCpuIndex++;
+                    continue;
                 }
 
-                if (m_mpi->UseGpuGdr() == 0)
+                //2. starting from 2nd loop : 2.1 async copy next 
+                m_gpuDataTransferers[gpuToCpuIndex]->CopyGPUToCPUAsync(gpuCopyBuffer->Data(), gpuCopyBuffer->GetNumElements(), m_intermediateCPUBuffers[gpuToCpuIndex].get());
+
+                //2.2 sync Allreduce
+                if (allReduceIndex != 0)
+                {
+                    m_gpuDataTransferers[allReduceIndex]->WaitForCopyGPUToCPUAsync();
+                }
+                reductionBuffer = m_intermediateCPUBuffers[allReduceIndex].get();
+                m_mpi->AllReduce(reductionBuffer, (currentGradientIndex == -1) ? m_aggregationBuffer->GetNumElements() : gradients[currentGradientIndex]->GetNumElements());
+
+                //2.3 Async h_to_g copy
+                cpuToGpuIndex = allReduceIndex;
+                m_gpuDataTransferers[cpuToGpuIndex]->CopyCPUToGPUAsync(m_intermediateCPUBuffers[cpuToGpuIndex].get(),
+                    (currentGradientIndex == -1) ? m_aggregationBuffer->GetNumElements() : gradients[currentGradientIndex]->GetNumElements(),
+                    (currentGradientIndex == -1) ? m_aggregationBuffer->Data() : gradients[currentGradientIndex]->Data());
+                allReduceIndex ++;
+                gpuToCpuIndex++;
+                currentGradientIndex = i;
+            }
+            // If currentGradientIndex == -2, it means no gradient index from m_gradientIndexToAggregate
+            if (currentGradientIndex == -2)
```

Comment:
Sorry, it's my bug. It should be "if (currentGradientIndex != -2)" in line 366. It is intended to handle no gradients case (if possible). Fixed in commit ada7743829834cbec63d5744962c41f70189dfe7